dial_opts.go

```diff
@@ -5,56 +5,108 @@ import (
 	VerifyCA = false
 )
 
-// WithSystemCerts is a dial option for requiring TLS with the system
-// certificate pool.
+// WithSystemCerts return a grpc.DialOption for requiring TLS that is
+// authenticated using the system-provided certificate authority chain.
 //
-// This function panics if the system pool cannot be loaded.
-func WithSystemCerts(insecureSkipVerify bool) grpc.DialOption {
+// If one cannot be found, this falls back to using a vendored version of
+// Mozilla's collection of root certificate authorities.
+func WithSystemCerts(insecureSkipVerify bool) (grpc.DialOption, error) {
 	certPool, err := x509.SystemCertPool()
 	if err != nil {
 		// Fall back to Mozilla collection of root CAs.
 		certPool, err = gocertifi.CACerts()
 		if err != nil {
 			// This library promises that this should never occur.
-			panic("gocertifi returned an error: " + err.Error())
+			return nil, fmt.Errorf("gocertifi returned an error: %w", err)
 		}
 	}
 
 	return grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
 		RootCAs:            certPool,
 		InsecureSkipVerify: insecureSkipVerify,
-	}))
+	})), nil
 }
 
-// WithCustomCerts is a dial option for requiring TLS from a specified path.
-// If the path is a directory, all certs are loaded. If it is an individual
-// file only the directly specified cert is loaded.
+func dirContents(dirPath string) ([][]byte, error) {
+	var allContents [][]byte
+	dirFS := os.DirFS(dirPath)
+	if err := fs.WalkDir(dirFS, ".", func(path string, d fs.DirEntry, err error) error {
+		if !d.IsDir() {
+			contents, err := fs.ReadFile(dirFS, d.Name())
+			if err != nil {
+				return err
+			}
+			allContents = append(allContents, contents)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return allContents, nil
+}
+
+// WithCustomCerts returns a grpc.DialOption for requiring TLS that is
+// authenticated using a certificate authority chain provided as a path on disk.
 //
-// This function panics if custom certificate pool cannot be instantiated.
-func WithCustomCerts(certPath string, insecureSkipVerify bool) grpc.DialOption {
-	certPool, err := customCertPool(certPath)
+// If the path is a directory, all files are loaded.
+func WithCustomCerts(certPath string, insecureSkipVerify bool) (grpc.DialOption, error) {
+	fi, err := os.Stat(certPath)
 	if err != nil {
-		panic(err)
+		return nil, err
+	}
+
+	var caFiles [][]byte
+	if fi.IsDir() {
+		caFiles, err = dirContents(certPath)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		contents, err := os.ReadFile(certPath)
+		if err != nil {
+			return nil, err
+		}
+		caFiles = append(caFiles, contents)
+	}
+
+	return WithCustomCertBytes(caFiles, insecureSkipVerify)
+}
+
+// WithCustomCertBytes returns a grpc.DialOption for requiring TLS that is
+// authenticated using a certificate authority chain provided in bytes.
+func WithCustomCertBytes(certChain [][]byte, insecureSkipVerify bool) (grpc.DialOption, error) {
```

Comment:
Yeah, I'd definitely define a new type for `certChain` as an alias of `map[string][]byte` and make it clear it is a set of filenames and their contents

Reply:
I did something a little different to avoid confusion and made it a callback for each file contents

Reply:
But this still takes in a slice of byte slices, which still feels off to me